### PR TITLE
Astro.resolve

### LINKED ISSUE #310: 💡 RFC: Astro.imageAsset(...)
Given this file structure:

```
src/components/my-astro-component
 - /index.astro
 - /image.png
```

and this code in `src/components/my-astro-component/index.astro`:

```
[...]
<style>
  .avatar {
    background: url({ Astro.imageAsset("./image.png") });
  }
[...]
```

I would like the component's image asset copied into a folder in in the final build (similar to adding it to /public manually) and for the `{ Astro.imageAsset("./image.png") }` text to be replaced with the public asset's path: 
`/astro-build-assets/[hash*]-image.png`

* where "hash" is a unique identifier for the image, mapped from the source path to the hash (so multiple references to the same imageAsset result in the same final path)
* Ideally the hash would be something like a quick md5() of the source path, relative to the project's root, so it would result in the same output each build unless the source moved or was renamed

This would allow astro components that contain images to be reused, and even distributed as a package, without the extra step of manually copying the static assets into a path defined by the component every time the component updates.

(currently my assets folder structure looks like a mirror of my component folder structure, which feels like having a nice closet already, but then building another closet off in the garage just for all of my socks)

Other considerations:

* the string passed into could have the caveat that it's a hardcoded path (not from a variable or otherwise constructed) so Astro can just parse it out of the component styles rather than requiring an actual Astro.imageAsset() function
* The implementation would have to limit the source to specific image types so a malicious component doesn't copy executable code into your build.
* any time Astro.imageAsset() is called (especially/only? when the source path contains "node_modules"), output information in the console, perhaps even a warning, so the site owner is aware of the assets being copied over.

I was going to build this myself using .fetchContent as a tour guide of the code and just submit a surprise PR, but the fetchContent feature isn't ready yet. ;( In any case, I'd be happy to build this feature myself given the green light and a tiny bit of guidance 💜

### PR DESCRIPTION
**Pull request by matthewp**
Closes #310 and #1071

## Changes

This adds support for Astro.resolve as described in the #1071 RFC.

## Testing

Tests added.

## Docs

Docs added.

### CODE REVIEW

File: packages/astro/src/compiler/codegen/utils.ts [hunk 1]
```diff
@@ -18,3 +19,20 @@ export function isImportMetaDeclaration(declaration: VariableDeclarator, metaNam
...
+
+const warnableRelativeValues = new Set([
+  'img+src',
+  'a+href',
+  'script+src',
+  'link+href',
+  'source+srcset'
+]);
+
+const nonRelative = /^(?!(https?|\/))/;
```
Comment (natemoo-re) [inline packages/astro/src/compiler/codegen/utils.ts:31]: Don't forget `data`!

File: packages/astro/src/compiler/codegen/index.ts [hunk 2]
```diff
@@ -118,9 +112,12 @@ async function getAttributes(attrs: Attribute[], state: CodegenState, compileOpt
         }
         continue;
       }
-      case 'Text':
-        result[attr.name] = JSON.stringify(getTextFromAttribute(val));
+      case 'Text': {
+        let text = getTextFromAttribute(val);
+        warnIfRelativeStringLiteral(nodeName, attr, text);
```
Comment (natemoo-re) [inline packages/astro/src/compiler/codegen/index.ts:117]: It's worth calling out—if we're warning for every relative string literal, does the user have a way to say "Don't warn me, I know what I'm doing?"

Comment (natemoo-re) [reply packages/astro/src/compiler/codegen/index.ts:117]: If not... why wouldn't we just automatically run `Astro.resolve`?

**Review by natemoo-re — COMMENTED** [review]
Looking good! A few comments.

File: packages/astro/src/compiler/codegen/index.ts (hunk 2 above)
Comment (matthewp) [reply packages/astro/src/compiler/codegen/index.ts:117]: Great points! First, this should definitely be using the logger so users log level choice gets honored.

Also, I should probably soften the language a little so it doesn't come across like an error. I definitely feel the annoyance of getting a warning repeatedly for something that you did on purpose. Any suggestions here? Would a link to docs be less annoying?

Comment (matthewp) [reply packages/astro/src/compiler/codegen/index.ts]: Ok, I've updated this to use the logger so you can turn it off the same way you can turn off other warnings. I don't think we have a flag to set a specific log level so we should add that in another issue. A way to turn off specific warnings sounds like a nice add as well but I can't think of a great solution to that at the moment (maybe a config file or something)

File: packages/astro/src/compiler/codegen/utils.ts [hunk 3]
```diff
@@ -18,3 +21,21 @@ export function isImportMetaDeclaration(declaration: VariableDeclarator, metaNam
...
```
Comment (jonathantneal) [inline packages/astro/src/compiler/codegen/utils.ts]: The variable name here is `nonRelative`, but it appears designed to match _relative_  URLs. Is this correct? Should it be renamed?

Also, according to [RFC 3987 ](http://www.faqs.org/rfcs/rfc3987.html), a valid scheme matches `ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )`

Therefore, a check for non-relative URLs might be:

```suggestion
const nonRelative = /^([A-Za-z][+-.0-9A-Za-z]*:|\/)/;
```

_☝️ I ordered the matches by character code. That’s why `+` comes before `0-9`, and `0-9` comes before `A-Z`._

Comment (matthewp) [reply packages/astro/src/compiler/codegen/utils.ts]: This matches relative links 🤦‍♂️, I'll rename and take your suggestion

**Review by natemoo-re — APPROVED** [review]
LGTM!

**Pull request merged**